### PR TITLE
[BEAM-6703] Added a phrase-triggered Jenkins job to test a Direct runner with Java 11 runtime

### PR DESCRIPTION
**Pull request by mwalenia**
This jenkins job compiles test sources with JDK 1.8 and runs tests with JDK 11 in order to check compatibility of the Direct worker with newer runtime. The Java paths are taken from [Apache confluence](https://cwiki.apache.org/confluence/display/INFRA/JDK+Installation+Matrix)

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run Direct ValidatesRunner in Java 11

**Comment by mwalenia** [comment]
R: @lgajowy @kkucharc

File: .test-infra/jenkins/README.md [hunk 1]
```diff
@@ -57,6 +57,7 @@ Beam Jenkins overview page: [link](https://builds.apache.org/view/A-D/view/Beam/
...
 | beam_PostCommit_Java_ValidatesRunner_Dataflow | [cron](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/), [phrase](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_PR/) | `Run Dataflow ValidatesRunner` | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow) |
+| beam_PostCommit_Java_ValidatesRunner_Direct_Java11 | [phrase](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Direct_PR) | `Run Direct ValidatesRunner in Java 11` | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Direct_PR/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Direct_PR) |
```
Comment (lgajowy) [inline .test-infra/jenkins/README.md]: I think the build status link (the red/green badge) should show only cron job status not Phrase triggered's. At least it seems to be the case for other tests in this file.

Why not add both jobs (Phrase triggered & cron job)?

File: .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+ * limitations under the License.
+ */
+
+import CommonJobProperties as commonJobProperties
+import PhraseTriggeringPostCommitBuilder
+
+
+// This job runs the suite of ValidatesRunner tests against the Dataflow
+// runner.
+PhraseTriggeringPostCommitBuilder.postCommitJob('beam_PostCommit_Java_ValidatesRunner_Direct',
```
Comment (lgajowy) [inline .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy]: Right now, this will only run when triggered manually from a PR comment. Should we also add a cron job for this (scheduled eg. once a day)? Or maybe it's better to have this tested post commit?

@aaltay @pabloem and @reuvenlax  what do you think? 

Tldr for you: this job tests if Direct runner's ValidatesRunner suite built with Java8 can be run with Java11, as we discussed. It's purpose is to ensure that new features developed with Java8 will stay compatible when running with Java 11. Thanks to running this suite, we were able to identify and fix a problem with schemas: [BEAM-6705](https://issues.apache.org/jira/browse/BEAM-6705)

File: .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        'Run Direct ValidatesRunner in Java 11', 'Direct Runner ValidatesRunner Tests for Java 11', this) {
+
+    description('Runs the ValidatesRunner suite on the Direct runner.')
+
+    def JAVA_11_HOME = '/home/jenkins/tools/java/latest11'
+    def JAVA_8_HOME = '/home/jenkins/tools/java/latest1.8'
+    // Set common parameters. Sets a 3 hour timeout.
+    commonJobProperties.setTopLevelMainJobProperties(delegate, 'master', 300)
+    previousNames(/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/)
```
Comment (lgajowy) [inline .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy]: Is this line needed? This job didn't have any previous name before (it's brand new).

File: .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import CommonJobProperties as commonJobProperties
+import PhraseTriggeringPostCommitBuilder
+
+
+// This job runs the suite of ValidatesRunner tests against the Dataflow
```
Comment (lgajowy) [inline .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy]: It's not running on Dataflow runner but on Direct. Please update the comment accordingly.

File: .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy (hunk 2 above)
Comment (lgajowy) [inline .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy]: `beam_PostCommit_Java_ValidatesRunner_Direct` -> `beam_PostCommit_Java11_ValidatesRunner_Direct`

Wdyt? IMO the name lacks some distinction of the fact that it's Java11

File: .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+// runner.
+PhraseTriggeringPostCommitBuilder.postCommitJob('beam_PostCommit_Java_ValidatesRunner_Direct',
+        'Run Direct ValidatesRunner in Java 11', 'Direct Runner ValidatesRunner Tests for Java 11', this) {
+
+    description('Runs the ValidatesRunner suite on the Direct runner.')
```
Comment (lgajowy) [inline .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy]: This is not only a validates runner suite - it builds with Java8 and runs with Java 11. The description should reflect that.

File: .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+    def JAVA_11_HOME = '/home/jenkins/tools/java/latest11'
+    def JAVA_8_HOME = '/home/jenkins/tools/java/latest1.8'
+    // Set common parameters. Sets a 3 hour timeout.
+    commonJobProperties.setTopLevelMainJobProperties(delegate, 'master', 300)
```
Comment (lgajowy) [inline .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy]: I think it's a 5 hour timeout (300 minutes) - could you change this to 3 hours?

File: .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy (hunk 2 above)
Comment (aaltay) [reply .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy]: I think running a post commit would be good. To prevent regressions over time.

Comment (pabloem) [reply .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy]: Agree that we should run this continuously. It being DirectRunner shouldn't take too much time either.

**Comment by mwalenia** [comment]
Run seed job

**Comment by mwalenia** [comment]
Run seed job

File: .test-infra/jenkins/README.md [hunk 7]
```diff
@@ -57,6 +57,7 @@ Beam Jenkins overview page: [link](https://builds.apache.org/view/A-D/view/Beam/
...
 | beam_PostCommit_Java_ValidatesRunner_Dataflow | [cron](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/), [phrase](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_PR/) | `Run Dataflow ValidatesRunner` | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow) |
+| beam_PostCommit_Java11_ValidatesRunner_Direct_Java | [cron](https://builds.apache.org/job/beam_PostCommit_Java11_ValidatesRunner_Direct), [phrase](https://builds.apache.org/job/beam_PostCommit_Java11_ValidatesRunner_Direct_PR) | `Run Direct ValidatesRunner in Java 11` | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java11_ValidatesRunner_Direct/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java11_ValidatesRunner_Direct) |
```
Comment (lgajowy) [inline .test-infra/jenkins/README.md]: Could you remove "java" at the end of the name?

`beam_PostCommit_Java11_ValidatesRunner_Direct_Java` => `beam_PostCommit_Java11_ValidatesRunner_Direct`

File: .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+// This job runs the suite of ValidatesRunner tests using Java 11 against the Direct
+// runner compiled with Java 8.
+PostcommitJobBuilder.postCommitJob('beam_PostCommit_Java11_ValidatesRunner_Direct',
+        'Run Direct ValidatesRunner in Java 11', 'Direct Runner ValidatesRunner Tests for Java 11', this) {
+
+    description('Builds the Direct Runner with Java 8 and runs ValidatesRunner test suite in Java 11.')
+
+    def JAVA_11_HOME = '/home/jenkins/tools/java/latest11'
+    def JAVA_8_HOME = '/home/jenkins/tools/java/latest1.8'
+    // Set common parameters. Sets a 3 hour timeout.
```
Comment (lgajowy) [inline .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy]: Can we remove this comment? I'd say that everything it says can be read from the method signature/implementation. Besides that, it can become obsolete and this can be misguiding in the future.

File: .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    commonJobProperties.setTopLevelMainJobProperties(delegate, 'master', 180)
+
+    // Publish all test results to Jenkins
```
Comment (lgajowy) [inline .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy]: Same as above

File: .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    publishers {
+        archiveJunit('**/build/test-results/**/*.xml')
+    }
+
+    // Gradle goals for this job.
```
Comment (lgajowy) [inline .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy]: and here too

File: .test-infra/jenkins/README.md (hunk 7 above)
Comment (lgajowy) [reply .test-infra/jenkins/README.md]: Besides that, I think it's great that you added the job here. Honestly, I'd forget this. :)

File: .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+ * limitations under the License.
+ */
+
+import CommonJobProperties as commonJobProperties
+import PostcommitJobBuilder
+
+
+// This job runs the suite of ValidatesRunner tests using Java 11 against the Direct
+// runner compiled with Java 8.
+PostcommitJobBuilder.postCommitJob('beam_PostCommit_Java11_ValidatesRunner_Direct',
```
Comment (lgajowy) [inline .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy]: +1 to adding a Postcommit job for this 👍

File: .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy (hunk 2 above)
Comment (lgajowy) [reply .test-infra/jenkins/job_PostCommit_Java_ValidatesRunner_Direct_Java11.groovy]: I felt the same, just wasn't sure. Thanks for opinions!

**Comment by mwalenia** [comment]
@lgajowy Fixed the cosmetics, thanks for your suggestions!

**Comment by lgajowy** [comment]
LGTM. Please fix commit messages (missing jira references), squash unnecessary commits (if any) and I will merge. Thanks!

**Comment by lgajowy** [comment]
LGTM. 👍  Please fix commit messages (missing jira references), squash unnecessary commits (if any) and I will merge. Thanks!

**Comment by mwalenia** [comment]
@lgajowy I fixed the commit history. Can you merge it?

**Comment by pabloem** [comment]
Thanks!

**Comment by pabloem** [comment]
exciting : D
